docs/learning/frontend/Extensions.md

```diff
@@ -0,0 +1,201 @@
+# iTwin.js Extensions
+
+An iTwin.js Extension is a separate JavaScript module that can load on demand into an iTwin.js frontend application.
+The separate deliverable enables Extensions to provide extensibility of an iTwin.js application at runtime.
+Extensions have access to a limited set of iTwin.js functionality through `@itwin/core-extension` to enable seamless integration with the host app.
+
+## What can extensions do?
+
+Extensions can be used for many different purposes, such as:
+- Add new a new [tool](./Tools.md) or [decorator](./ViewDecorations.md) to an existing application to better support your custom workflows.
+- Write event based processing, i.e., subscribe to an iModel Event, or Unified Selection Event, and process that change.
+
+## How to get started
+
+An iTwin.js Extension is just a javascript file and a manifest (i.e., a package.json file with some additional properties).
+To get started, create a new directory for the Extension.
+
+### Setup the Manifest
+The first step to creating an extension is to create a manifest.
+Create the `package.json` by running, in the directory created above:
+```
+npm init --yes
+```
+The following properties must be added to the package.json file:
+- _Name_: the name of the Extension.
+- _Version_: the version of the Extension in the format _x.x.x_.
+- _Main_: where to find the javascript file.
+- _ActivationEvents_: events that define when the iTwin.js application should load and run your Extension. Currently, we only support `onStartup`, which will execute the Extension as soon as it is loaded.
```

Comment:
Yes I'll add some more details - onStartup will execute whenever the extension is added to the app (so whenever you call IModelApp.extensionAdmin.addExtension()). I guess that could be before a viewport is created. But for those based on the Viewer component, the viewport will be created beforehand anyways

Reply:
Thanks but can you - or someone explain *to me* when exactly `addExtension` gets invoked, particularly for a build-time extension? IModelApp.startup happens, then possibly a Viewport gets opened, then ...?

Reply:
It really depends on the application and when you choose to invoke addExtension. In the Viewer, for example, we invoke addExtension right after IModelApp.startup (so IModelApp.onAfterStartup is unnecessary). So my example in this doc probably wouldn't work as is in the viewer.
In hindsight, we probably should have waited for a viewport to be opened, but I think that could be defined as a separate activation event in the future.
Most use cases for Extensions don't need to wait for a viewport because they don't act right away (unlike my example in this doc) - some outside event triggers some interaction (for example, a postMessage event).

Reply:
None of that is clear to me from this documentation.

Reply:
I don't want to overcomplicate the tutorial... imo "execute the Extension as soon as it is added" is enough for now because it describes exactly what happens: extensions are executed in addExtension()
But you are right about the viewport stuff and we should expand activation events in the future when we actually have more of them and be more clear which events are good for which situations.

Reply:
If that's what you're going with please remove all the stuff that depends on viewports existing, because they won't. Registering tools and decorators (via ViewManager) will make better examples.

Reply:
Removing the isView3d query. Just the tool should be a good enough example anyway